### PR TITLE
C# test runner for Xamarin.Android tests

### PR DESCRIPTION
**Pull request by grendello**
`XAT` (for `Xamarin Android Testing`, imaginative) is a utility to replace our current MSBuild-based test infrastructure with a single command
that allows one to list all the tests, run all or any of them, easily select single (or more) tests from the test suites, no matter whether they're APK
or host unit tests. The intention is to make it easier for anyone not dabbling with our test suite day in day out to add, run, modify tests. The ideal
here is that one would only need to edit `tools/xat/TestCollection.cs` to do what they want.

Why is it needed? The current MSBuild architecture scatters the test running code, data describing tests across at 5 files in 4 different directories.
The `.targets` and `.projitems` files additionally reference each other causing confusion every time one (or maybe it's just me) wants to run, add or
modify any code to run and list test suites and individual tests. Additionally, wrapping everything in MSBuild causes the test run output to be 
voluminous and increases signal-to-noise ratio, making reading the output harder.

What needs reviewing the most:

 * tools/xat/TestsAPI/*
 * tools/xat/TestCollection.cs

To run the tool go to `tools/xat` and run the `run.sh` script on Unix
or, on Windows, run:

    msbuild /r
		..\..\bin\Debug\xat.exe

To see command line usage, pass the `--help` option, each sub-command
also excepts one.

What remains to be done:

 * Removal of the obsoleted MSBuild `targets` and `projitems` files (not doing it now to make review easier)
 * Implementation of test groups (to be able to say `-s NetworkTests` to run all the network tests from both APK and host test suites, for instance
 * Changes to `build-scripts/automation/` to use the new tool
 * General code polish
 * Support for .NET 6 tests

### CODE REVIEW

**Review by jonathanpeppers — COMMENTED** [review]
Main thing I was wondering is if there should be `.yaml` changes here that use the new `xat` tool?

Would it simplify things like:
* https://github.com/xamarin/xamarin-android/blob/master/build-tools/automation/yaml-templates/apk-instrumentation.yaml
* https://github.com/xamarin/xamarin-android/blob/master/build-tools/automation/yaml-templates/run-msbuild-device-tests.yaml

**Comment by grendello** [comment]
> Main thing I was wondering is if there should be `.yaml` changes here that use the new `xat` tool?
> 
> Would it simplify things like:
> 
>     * https://github.com/xamarin/xamarin-android/blob/master/build-tools/automation/yaml-templates/apk-instrumentation.yaml
> 
>     * https://github.com/xamarin/xamarin-android/blob/master/build-tools/automation/yaml-templates/run-msbuild-device-tests.yaml

It's on the TODO list (see the bottom of the PR message). I didn't want to muddle the PR with these changes now - the first thing is to get it working right, integration will come later. For the moment, it's possible to test it from command line.

File: tools/xat/xat.targets [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+    <ParameterGroup>
+      <SourceFile ParameterType="Microsoft.Build.Framework.ITaskItem" Required="true" />
+      <DestinationFile ParameterType="Microsoft.Build.Framework.ITaskItem" Required="true" />
+      <Replacements ParameterType="System.String[]" />
+      <ReplacementFilePath ParameterType="System.String" />
+    </ParameterGroup>
+    <Task>
+      <Code Type="Class" Language="cs" Source="$(ReplaceFileContentSource)"/>
+    </Task>
+  </UsingTask>
```
[the remark this answers is not in the document]
Comment (grendello) [reply tools/xat/xat.targets:27]: Yeah, it should be possible. I copied the code from `xaprepare` without much thinking.

File: build-tools/xa-prep-tasks/Xamarin.Android.BuildTools.PrepTasks/ProcessLogcatTiming.Shared.cs [hunk 2]
```diff
@@ -0,0 +1,121 @@
+//
```
Comment (jonpryor) [inline build-tools/xa-prep-tasks/Xamarin.Android.BuildTools.PrepTasks/ProcessLogcatTiming.Shared.cs:1]: …odd that this file wasn't treated as a file rename from `ProcessLogcatTiming.cs`…

**Comment by jonpryor** [comment]
Commit message mistake?

> To run the tool go to `tools/xat` and run the `run.sh` script on Unix

That doesn't work:

```bash
% ./tools/xat/run.sh 
Microsoft (R) Build Engine version 16.6.0 for Mono
Copyright (C) Microsoft Corporation. All rights reserved.

MSBUILD : error MSB1011: Specify which project or solution file to use because this folder contains more than one project or solution file.
```

This is presumably due to [line 2 in `run.sh`](https://github.com/xamarin/xamarin-android/pull/5369/files#diff-f8eb0c911d0f677cb8728bc3149e5bc3a8be29be6a8e43e00a045726de190ac7R2):

```bash
msbuild /r
```

If I edit `run.sh` to remove the `msbuild /r`, it doesn't work correctly if I run from "topdir":

```
% ./tools/xat/run.sh --help
Cannot open assembly '../../bin/Debug/bin/xat.exe': No such file or directory.
```

`run.sh` should presumably be:

```bash
#!/bin/bash -e

HERE="$(cd "$(dirname "$0")" && pwd)"

exec mono --debug "$HERE/../../bin/Debug/bin/xat.exe" "$@"
```

which *does* execute for me.

**Comment by jonpryor** [comment]
Random musings as play around with `tools/xat/run.sh`…

I would suggest attempting to make the `tools/xat/run.sh --help` output less anemic/more useful:

```
% tools/xat/run.sh --help

Locating essential tool binaries
  • git
     Found: /usr/local/bin/git
  • homebrew
     Found: /usr/local/bin/brew
  • pkgutil
     Found: /usr/sbin/pkgutil
usage: xat COMMAND [OPTIONS]

Xamarin.Android v11.2.99 test shell

Global options:
  -v, --verbosity=LEVEL      Set console log verbosity to LEVEL. Level name may
                               be abbreviated to the smallest unique part (one
                               of: silent, quiet, normal, verbose, diagnostic).
                               Default: normal
  -c, --configuration=CONFIGURATION
                             Set build CONFIGURATION instead of the default '
                               Debug'
  -D, --dull-mode            Run xat in "dull mode" - no colors, no emoji
  -E, --no-emoji             Run xat without using emoji, but still allow color

Available commands:
        list                 List known tests/categories/etc
        run                  Invoke selected tests/categories
```

I want to read top-down, assuming that information at the top is "more important".  (See also: every `man` page?)

Yet here we have a `Locating essential tool binaries` *section*, which *isn't* useful.  Then we get to a brief description, but nothing really useful.

(Actually, *every command* prints out this `Locating essential tool binaries` section, which is *useless noise*.  Unless something goes wrong, why do I want to see this?  Why should `xat list` show me the path to `git` or `brew`?)

I would suggest output similar to:

```
% run.sh --help
usage: xat COMMAND [OPTIONS]

Runs xamarin-android unit tests and applications.

See tools/xat/README.md for more information.

Global options:
  -v, --verbosity=LEVEL      Set console log verbosity to LEVEL. Level name may
                               be abbreviated to the smallest unique part (one
                               of: silent, quiet, normal, verbose, diagnostic).
                               Default: normal
  -c, --configuration=CONFIGURATION
                             Set build CONFIGURATION instead of the default '
                               Debug'
  -D, --dull-mode            Run xat in "dull mode" - no colors, no emoji
  -E, --no-emoji             Run xat without using emoji, but still allow color

Available commands:
        list                 List known tests/categories/etc
        run                  Invoke selected tests/categories

Examples:

  …

```

A month+ later, I'm still not able to remember, understand, or conceptualize the difference between a "group" and a "suite".  Perhaps different nouns should be used?

I can't quickly see what `xat list -m` *does*, other than emit more blank lines.

I wonder if a concept of "tags" would be more useful?  E.g. `AOT_Mono_Android` could have tags `{ aot Mono.Android.dll }`, `AOT_Xamarin_Forms_Performance_Integration` could have tags of `{ aot forms timing }`, etc.?

Is there/should there be a way to *search*/*filter* `xat list` output? `xat list -g`/`xat list -s` does a form of filtering, but -- as per above -- I don't *understand* this difference.

***EXAMPLES***.  The "Documentation" -- both in the commit message, and `xat --help`, and `xat run --help`, *everything* needs examples.

I should have read `tools/xat/README.md` first.  Doh.  It still lacks examples.

"test" is possibly an overused noun.  Consider the `xat run --help` output:

```
% tools/xat/run.sh run --help
…
usage: xat run [OPTIONS]
By default the command runs all the tests.
```

OK, it runs tests.

```
  -g, --group=GROUPS         Run the specified GROUPS. Takes a comma-separated
                               list of test group names or can be repeated on
                               command line
  -s, --suite=SUITES         Run the specified SUITES. Takes a comma-separated
                               list of test suite names or can be repeated on
                               command line
```

what's the relation between these, again?

```
The arguments below use suite-specific format for each entry. 
```

Oh boy.

```
… They are ignored if a test suite runs as part of predefined group:
```

I have no idea what this means. What makes a "group" a "predefined" group? Why would they be ignored?  Why should they be ignored?  If I do `xat run -g APK_all_tests`, shouldn't `-n`/`-d`/`-a` be applicable?  This header suggests ¯\_(ツ)_/¯.

```
  -t, --test=TESTS           Run the specified TESTS. Takes a comma-separated
                               list of test names
```

So run the specified tests.  What happens with `-t` is used in combination with `-g=` or `-s=`? Is it additive? Set intersection?  Other?

```
  -i, --include-categories=CATEGORIES
                             Run tests in the specified CATEGORIES. Takes a
                               comma-separated list of test categories
```

So tests have categories?  (This is presumably the NUnit/xUnit concept of categories?) How does this relate to groups & suites?  *Does* it?  *Should* it?

What *is* `Unit_No_Node`? It doesn't appear to *do* anything:

```
% tools/xat/run.sh run -g Unit_No_Node
… prints out tool paths …
```

Some of the suite names also contain "duplicative" strings, and not necessarily in a way that "conveys meaning":

  * `JavaInterop_Java_Interop_Tests`
  * `JavaInterop_Java_Interop_Dynamic_Tests`
  * `JavaInterop_Java_Interop_Export_Tests`
  * …

Both "Java" and "Interop" are repeated twice in each suite name, but what's it *mean*? It needs to be unique, certainly, but why not trim off the `JavaInterop_` prefix?  (Perhaps another place where "tags" could be a useful concept?)

I suggested earlier that "names" -- not sure if I meant group name or suite name, and I still don't know the difference -- could/should be *filesystem paths*, because then at least I *know* what it corresponds to.  I have no idea what `JavaInterop_Java_Interop_Tests` is.  There are *at least two separate projects* for "`Java.Interop-Tests`":

  * `external/Java.Interop/tests/Java.Interop-Tests`: Desktop unit tests for `Java.Interop.dll` (in `Java.Interop-Tests.dll`)
  * `tests/Mono.Android-Tests/Java.Interop-Tests`: Unit tests for `Java.Interop.dll` for use within the `Mono.Android-Tests`.

With some thought, I *think* the second one isn't "applicable", because it'll be "included" in `Mono.Android-Tests`, but this is *thinking* that I don't necessarily want to perform. Reading the source code doesn't immediately help either!

```csharp
{ "Java.Interop-Tests.dll",                            Path.Combine ("Java.Interop-Tests",                            "Java.Interop-Tests.csproj") },
```

How many `Java.Interop-Tests/Java.Interop-Tests.csproj` files are there?

```
% find . -iname Java.Interop-Tests.csproj
./tests/Mono.Android-Tests/Java.Interop-Tests/Java.Interop-Tests.csproj
./external/Java.Interop/tests/Java.Interop-Tests/Java.Interop-Tests.csproj
```

That's right, 2.

In short, the `JavaInterop_Java_Interop_Tests` suite name does me no favors, the "Host Unit" name of `JavaInterop: Java.Interop-Tests` doesn't help me, and the *source code* doesn't help me.  The only thing that *might* help, maybe, *if I got it right*, is mental effort to *exclude* one of the tests projects I know of, on the assumption that it's inclusion "doesn't make sense."

I'm thinking too hard already. :-(

What *would* be a useful "grouping" to me would be "all AOT tests".  No such grouping exists.

```
% ./tools/xat/run.sh run -s Mono_Android
…
```

What's the output look like when there's a *failure*?  Examples welcome!

As-is, output is, again, *verbose* (yay?), but not all is necessarily *useful*?

```
% ./tools/xat/run.sh run -s Mono_Android

Locating essential tool binaries
  • git
     Found: /usr/local/bin/git
  • homebrew
     Found: /usr/local/bin/brew
  • pkgutil
     Found: /usr/sbin/pkgutil
Main log file: …/xamarin-android/bin/TestDebug/tests-20210115T210001.log
Tool paths:
  • ADB: /Users/jon/android-toolchain/sdk/platform-tools/adb
  • AVD Manager: /Users/jon/android-toolchain/sdk/cmdline-tools/1.0/bin/avdmanager
  • BundleTool: …/xamarin-android/bin/Debug/lib/xamarin.android/xbuild/Xamarin/Android/bundletool.jar
  • Emulator: /Users/jon/android-toolchain/sdk/emulator/emulator
  • Java: /Users/jon/android-toolchain/jdk-11/bin/java
  • NUnit: …/xamarin-android/packages/nunit.consolerunner/3.11.1/tools/nunit3-console.exe

Running suite: Mono.Android
Running global initialization steps for suite kind 'APK', family 'APK'
 | 29
 | package:/system/priv-app/Shell/Shell.apk
 | List of devices attached
 | 98211FFBA006ND	device
 | 
Detected ADB target: 98211FFBA006ND
Building suite 'Mono.Android'
xabuild "/p:Configuration=Debug" /v:normal "/bl:../../bin/BuildDebug/msbuild-20210115T210001-build_test_Mono_Android.binlog" "/r" "/t:SignAndroidPackage" "Mono.Android-Tests.csproj" 
[MSBuild] running
[MSBuild] log file: bin/TestDebug/tests-20210115T210001.msbuild.build_test_Mono_Android.log

 | Success
Deploying package: Mono.Android_Tests-Signed.apk
 | Performing Streamed Install
 | Success
Running instrumentation tests
Starting test run for:
  Package: Mono.Android_Tests
  Instrumentation: Mono.Android_Tests/xamarin.android.runtimetests.NUnitInstrumentation
 | INSTRUMENTATION_RESULT: failed=0
 | INSTRUMENTATION_RESULT: filtered=0
 | INSTRUMENTATION_RESULT: inconclusive=0
 | INSTRUMENTATION_RESULT: nunit2-results-path=/storage/emulated/0/Android/data/Mono.Android_Tests/files/Documents/TestResults.NUnit.xml
 | INSTRUMENTATION_RESULT: passed=871
 | INSTRUMENTATION_RESULT: run=873
 | INSTRUMENTATION_RESULT: skipped=2
 | INSTRUMENTATION_RESULT: total=0
 | INSTRUMENTATION_CODE: -1
Logcat output path: …/xamarin-android/bin/TestDebug/logcat-Debug-Mono.Android_Tests.txt
Test results output path: …/xamarin-android/bin/TestDebug/TestResult-Mono.Android_Tests.xml
 | /storage/emulated/0/Android/data/Mono.Android_Tests/files/Documents/TestResults.NUnit.xml: 1 file pulled, 0 skipped. 27.7 MB/s (223350 bytes in 0.008s)
Processing timing results
   Results path: …/xamarin-android/TestResult-Mono.Android_Tests-times.csv
   Definitions path: …/xamarin-android/build-tools/scripts/TimingDefinitions.txt
Renaming test cases for 'Mono.Android': instrumentation 'xamarin.android.runtimetests.NUnitInstrumentation'
File fixed up: …/xamarin-android/TestResult-Mono.Android_Tests-Debug.xml
Running global shutdown steps for suite kind 'APK, family 'APK'

Suite execution times:
  • Mono_Android (APK): 00:00:27.1318713
Test suite succeeded
Main log file: …/xamarin-android/bin/TestDebug/tests-20210115T210001.log
```

`Locating essential tool binaries` isn't helpful. `Tool paths` isn't helpful, and tool invocations will be in the generated log files, so if they are needed, they can be obtained.


```
Running global initialization steps for suite kind 'APK', family 'APK'
 | 29
 | package:/system/priv-app/Shell/Shell.apk
 | List of devices attached
 | 98211FFBA006ND	device
```

I have no idea what this means. Is this *needed* for console output, and not log file output?

```
Building suite 'Mono.Android'
xabuild "/p:Configuration=Debug" /v:normal "/bl:../../bin/BuildDebug/msbuild-20210115T210001-build_test_Mono_Android.binlog" "/r" "/t:SignAndroidPackage" "Mono.Android-Tests.csproj" 
[MSBuild] running
[MSBuild] log file: bin/TestDebug/tests-20210115T210001.msbuild.build_test_Mono_Android.log

 | Success
```

Yay, success!  All useful!

```
Starting test run for:
  Package: Mono.Android_Tests
  Instrumentation: Mono.Android_Tests/xamarin.android.runtimetests.NUnitInstrumentation
 | INSTRUMENTATION_RESULT: failed=0
 | INSTRUMENTATION_RESULT: filtered=0
 | INSTRUMENTATION_RESULT: inconclusive=0
 | INSTRUMENTATION_RESULT: nunit2-results-path=/storage/emulated/0/Android/data/Mono.Android_Tests/files/Documents/TestResults.NUnit.xml
 | INSTRUMENTATION_RESULT: passed=871
 | INSTRUMENTATION_RESULT: run=873
 | INSTRUMENTATION_RESULT: skipped=2
 | INSTRUMENTATION_RESULT: total=0
 | INSTRUMENTATION_CODE: -1
Logcat output path: /Volumes/Xamarin-Work/xamarin-android/bin/TestDebug/logcat-Debug-Mono.Android_Tests.txt
Test results output path: /Volumes/Xamarin-Work/xamarin-android/bin/TestDebug/TestResult-Mono.Android_Tests.xml
 | /storage/emulated/0/Android/data/Mono.Android_Tests/files/Documents/TestResults.NUnit.xml: 1 file pulled, 0 skipped. 27.7 MB/s (223350 bytes in 0.008s)
```

Notably absent from this section: "Success", or any "clear" indicator that it was successful, *unless* you *know* that `INSTRUMENTATION_CODE: -1` means "👍". An explicit "Success" message might be nice here?


```
Processing timing results
   Results path: /Volumes/Xamarin-Work/xamarin-android/TestResult-Mono.Android_Tests-times.csv
   Definitions path: /Volumes/Xamarin-Work/xamarin-android/build-tools/scripts/TimingDefinitions.txt
Renaming test cases for 'Mono.Android': instrumentation 'xamarin.android.runtimetests.NUnitInstrumentation'
File fixed up: /Volumes/Xamarin-Work/xamarin-android/TestResult-Mono.Android_Tests-Debug.xml
Running global shutdown steps for suite kind 'APK, family 'APK'
```

Not sure if it's useful for console output.  `Main log file` contains all the same information and more; perhaps "Processing timing results" doesn't *need* to be in the console output?

File: build-tools/xa-prep-tasks/Xamarin.Android.BuildTools.PrepTasks/ProcessLogcatTiming.Shared.cs (hunk 2 above)
Comment (grendello) [reply build-tools/xa-prep-tasks/Xamarin.Android.BuildTools.PrepTasks/ProcessLogcatTiming.Shared.cs:1]: :shrug:

**Comment by grendello** [comment]
> A month+ later, I'm still not able to remember, understand, or conceptualize the difference between a "group" and a "suite". Perhaps different nouns should be used?

 * The smallest unit is a single `test` (unit, integration, performance etc)
 * Individual tests are gathered in `suites` (e.g. `Mono.Android Tests`, `BCL tests`, `XF Integration tests`) which translate to assemblies with tests, apks with tests, eventually native apps with tests
 * The test suites are **grouped** together to (primarily) reflect the groupings used by our Azure pipelines. The groups are **arbitrary** collections - defined by us in order to make CI runs more finely grained (and remove some YAML code, as a side effect)

We can come up with a number of synonyms for all three nouns, but what would be the point? Would a `suite collection` or `suite set` be more understandable than `suite group`? Commonly, we refer to our **suites** collectively as "tests", but since a "test" is a single, smallest, unit of a **test suite** we can't call `suites` "tests" for the purpose of this utility. After all, sometimes you will want to run an individual test from a test suite, and referring to it as "running a test in a test" would be rather unclear. On the other hand, running a "test in suite X" or a  "group of test suites named X" is quite clear IMO.

> I can't quickly see what xat list -m does, other than emit more blank lines.

I'm not sure which blank lines are you referring to? The output below clearly shows division into test suites and their grouping ("groups") and the empty lines are to separate the individual suites/groups so that one isn't faced with a wall of text:
```shell
$ ./run.sh list -m

Suites:
 • AOT_Mono_Android (APK: 'AOT: Mono.Android')
 • AOT_Xamarin_Android_Locale (APK: 'AOT: Xamarin.Android Locale')
 • AOT_Xamarin_Forms_Performance_Integration (APK: 'AOT: Xamarin.Forms Performance Integration')
 • AppBundle_Mono_Android (APK: 'AppBundle: Mono.Android')
 • BCL (APK: 'BCL')
 • Build_Tests (Host Unit: 'Build Tests')
 • Embedded_DSO (APK: 'Embedded DSO')
 • JavaInterop_generator_Tests (Host Unit: 'JavaInterop: generator-Tests')
 • JavaInterop_Java_Interop_PerformanceTests (Host Unit: 'JavaInterop: Java.Interop-PerformanceTests')
 • JavaInterop_Java_Interop_Tests (Host Unit: 'JavaInterop: Java.Interop-Tests')
 • JavaInterop_Java_Interop_Dynamic_Tests (Host Unit: 'JavaInterop: Java.Interop.Dynamic-Tests')
 • JavaInterop_Java_Interop_Export_Tests (Host Unit: 'JavaInterop: Java.Interop.Export-Tests')
 • JavaInterop_Java_Interop_Tools_Generator_Tests (Host Unit: 'JavaInterop: Java.Interop.Tools.Generator-Tests')
 • JavaInterop_Java_Interop_Tools_JavaCallableWrappers_Tests (Host Unit: 'JavaInterop: Java.Interop.Tools.JavaCallableWrappers-Tests')
 • JavaInterop_Java_Interop_Tools_JavaSource_Tests (Host Unit: 'JavaInterop: Java.Interop.Tools.JavaSource-Tests')
 • JavaInterop_logcat_parse_Tests (Host Unit: 'JavaInterop: logcat-parse-Tests')
 • JavaInterop_Xamarin_Android_Tools_ApiXmlAdjuster_Tests (Host Unit: 'JavaInterop: Xamarin.Android.Tools.ApiXmlAdjuster-Tests')
 • JavaInterop_Xamarin_Android_Tools_Bytecode_Tests (Host Unit: 'JavaInterop: Xamarin.Android.Tools.Bytecode-Tests')
 • JavaInterop_Xamarin_SourceWriter_Tests (Host Unit: 'JavaInterop: Xamarin.SourceWriter-Tests')
 • JcwGen (APK: 'JcwGen')
 • Mono_Android (APK: 'Mono.Android')
 • MonoBundle_Mono_Android (APK: 'MonoBundle: Mono.Android')
 • MonoBundle_Xamarin_Forms_Performance_Integration (APK: 'MonoBundle: Xamarin.Forms Performance Integration')
 • MSBuild_Device_Integration (Host Unit: 'MSBuild Device Integration')
 • MSBuildTiming_HelloWorld (MSBuild Timing: 'MSBuildTiming: HelloWorld')
 • MSBuildTiming_Xamarin_Forms_Integration (MSBuild Timing: 'MSBuildTiming: Xamarin.Forms-Integration')
 • MultiDex_Mono_Android (APK: 'MultiDex: Mono.Android')
 • Xamarin_Android_Locale (APK: 'Xamarin.Android Locale')
 • Xamarin_Forms_Performance_Integration (APK: 'Xamarin.Forms Performance Integration')

Groups:
 • APK_all_tests: 'APK: all tests'
   • AOT_Mono_Android (APK: 'AOT: Mono.Android')
   • AOT_Xamarin_Android_Locale (APK: 'AOT: Xamarin.Android Locale')
   • AOT_Xamarin_Forms_Performance_Integration (APK: 'AOT: Xamarin.Forms Performance Integration')
   • AppBundle_Mono_Android (APK: 'AppBundle: Mono.Android')
   • BCL (APK: 'BCL')
   • Embedded_DSO (APK: 'Embedded DSO')
   • JcwGen (APK: 'JcwGen')
   • MonoBundle_Mono_Android (APK: 'MonoBundle: Mono.Android')
   • MonoBundle_Xamarin_Forms_Performance_Integration (APK: 'MonoBundle: Xamarin.Forms Performance Integration')
   • Mono_Android (APK: 'Mono.Android')
   • MultiDex_Mono_Android (APK: 'MultiDex: Mono.Android')
   • Xamarin_Android_Locale (APK: 'Xamarin.Android Locale')
   • Xamarin_Forms_Performance_Integration (APK: 'Xamarin.Forms Performance Integration')

 • Unit_Java_Interop: 'Unit: Java.Interop'
   • JavaInterop_Java_Interop_Dynamic_Tests (Host Unit: 'JavaInterop: Java.Interop.Dynamic-Tests')
   • JavaInterop_Java_Interop_Export_Tests (Host Unit: 'JavaInterop: Java.Interop.Export-Tests')
   • JavaInterop_Java_Interop_PerformanceTests (Host Unit: 'JavaInterop: Java.Interop-PerformanceTests')
   • JavaInterop_Java_Interop_Tests (Host Unit: 'JavaInterop: Java.Interop-Tests')
   • JavaInterop_Java_Interop_Tools_Generator_Tests (Host Unit: 'JavaInterop: Java.Interop.Tools.Generator-Tests')
   • JavaInterop_Java_Interop_Tools_JavaCallableWrappers_Tests (Host Unit: 'JavaInterop: Java.Interop.Tools.JavaCallableWrappers-Tests')
   • JavaInterop_Java_Interop_Tools_JavaSource_Tests (Host Unit: 'JavaInterop: Java.Interop.Tools.JavaSource-Tests')
   • JavaInterop_Xamarin_Android_Tools_ApiXmlAdjuster_Tests (Host Unit: 'JavaInterop: Xamarin.Android.Tools.ApiXmlAdjuster-Tests')
   • JavaInterop_Xamarin_Android_Tools_Bytecode_Tests (Host Unit: 'JavaInterop: Xamarin.Android.Tools.Bytecode-Tests')
   • JavaInterop_Xamarin_SourceWriter_Tests (Host Unit: 'JavaInterop: Xamarin.SourceWriter-Tests')
   • JavaInterop_generator_Tests (Host Unit: 'JavaInterop: generator-Tests')
   • JavaInterop_logcat_parse_Tests (Host Unit: 'JavaInterop: logcat-parse-Tests')

 • Unit_MSBuild_Extra_Tests_No_Node: 'Unit: MSBuild Extra Tests No Node'
   • Build_Tests (Host Unit: 'Build Tests')

 • Unit_MSBuild_Legacy_macOS_Node_1: 'Unit: MSBuild Legacy - macOS Node 1'
   • Build_Tests (Host Unit: 'Build Tests')

 • Unit_MSBuild_Legacy_macOS_Node_2: 'Unit: MSBuild Legacy - macOS Node 2'
   • Build_Tests (Host Unit: 'Build Tests')

 • Unit_MSBuild_Legacy_macOS_Node_3: 'Unit: MSBuild Legacy - macOS Node 3'
   • Build_Tests (Host Unit: 'Build Tests')

 • Unit_MSBuild_Smoke_Tests: 'Unit: MSBuild Smoke Tests'
   • Build_Tests (Host Unit: 'Build Tests')
   • MSBuild_Device_Integration (Host Unit: 'MSBuild Device Integration')

 • Unit_No_Node: 'Unit: No Node'

 • Unit_On_Device_macOS: 'Unit: On Device macOS'

 • Unit_TimeZone_Node_1: 'Unit: TimeZone Node 1'

 • Unit_TimeZone_Node_2: 'Unit: TimeZone Node 2'

 • Unit_TimeZone_Node_3: 'Unit: TimeZone Node 3'
```

> I wonder if a concept of "tags" would be more useful? E.g. AOT_Mono_Android could have tags { aot Mono.Android.dll }, AOT_Xamarin_Forms_Performance_Integration could have tags of { aot forms timing }, etc.?

I don't know. That's why I asked for a review from everybody involved in test suites. The goal is not to make it (primarily) pretty but to make it possible to run groups/tags/collections/sets/gatherings of `test suites` comfortably from both command line locally and the CI.  
As it stands now, locally we're unable to run the "groups" of `test suites` locally in the same manner as we run them as part of the CI pipeline.  If people prefer "tags" instead of "groups", then I don't have a preference - the mechanics will be exactly the same, what will
change is the terminology and names of "groups".

> Is there/should there be a way to search/filter xat list output? xat list -g/xat list -s does a form of filtering, but -- as per above -- I don't understand this difference.

Filtering based on what criteria? Name? Test categories (note they are different in different suites)? Please remember that we are dealing with a **really** diverse set of test **kinds**:

 * NUnit unit tests running on host
 * NUnit unit tests running on devices
 * xUnit unit tests running on devices
 * Integration tests running on host
 * Integration tests running on devices
 * (in the future) native tests running on devices
 * (in the future) native tests running on host

*All** of them differe (even if slightly) in their terminology, the way they categorize tests, the way they discover tests etc. `XAT` attempts to gather them under a single umbrella, using coherent terminology which, by necessity, has to be wide/vague enough to encompass all
the terminologies used throughout the suites.

> What would be a useful "grouping" to me would be "all AOT tests". No such grouping exists.

Yes, it doesn't exist - this is **precisely** the input for which I asked the reviewers: how to group the tests and which ones. I added some "sample" groups based on what is used in our CI pipeline YAML files, and that is what you see today.

>> The arguments below use suite-specific format for each entry. 

> Oh boy.

>> … They are ignored if a test suite runs as part of predefined group:

> I have no idea what this means. What makes a "group" a "predefined" group? Why would they be ignored? Why should they be ignored? If I do xat run -g APK_all_tests, shouldn't -n/-d/-a be applicable? This header suggests ¯_(ツ)_/¯.

What makes it "predefined" is the hardcoding of the group in the source code. They would be ignored because if your group contains both `xUnit` tests and `integration tests` using **DIFFERENT** definitions of categories/filters/whatever and **each** of them using a different format
for selecting tests/test classes/whatever - how do you expect to specify the filters so that the right format gets to the right test driver? Sure, we could add suboptions to suboptions and complicate the command line to Nth degree, but that would be rather useless. Let me reiterate:
groups exist for the purpose of running **sets** (groups, collections, gatherings) of **test suites** in a shortcut way **primarily** on CI. They exist so that you can put `xat -g SomeGroup` in the YAML definition and be done with it.

> So run the specified tests. What happens with -t is used in combination with -g= or -s=? Is it additive? Set intersection? Other?

As clearly stated below:

```shell
$ ./run.sh run --help
usage: xat run [OPTIONS]
By default the command runs all the tests. Options can be combined and repeated
on the command line unless stated otherwise

  -m, --msbuild=BINARY       Use the specified MSBuild BINARY. If omitted, XAT
                               will first try to find `xabuild` and fall back
                               to `msbuild`

  -g, --group=GROUPS         Run the specified GROUPS. Takes a comma-separated
                               list of test group names or can be repeated on
                               command line
  -s, --suite=SUITES         Run the specified SUITES. Takes a comma-separated
                               list of test suite names or can be repeated on
                               command line

The arguments below use suite-specific format for each entry. They are ignored
if a test suite runs as part of predefined group:
  -t, --test=TESTS           Run the specified TESTS. Takes a comma-separated
                               list of test names
  -i, --include-categories=CATEGORIES
                             Run tests in the specified CATEGORIES. Takes a
                               comma-separated list of test categories
  -e, --exclude-categories=CATEGORIES
                             Do not tun tests in the specified CATEGORIES.
                               Takes a comma-separated list of test categories
  -j, --include-tests=TESTS  Run the specified TESTS. Takes a comma-separated
                               list of test names
  -f, --exclude-tests=TESTS  Do not run the specified TESTS. Takes a comma-
                               separated list of test names

  -n, --new-emulator         Create emulator AVD even if it already exists
  -d, --device=DEVICE        ID/name of Android DEVICE to use
  -a, --adb-options=OPTIONS  Additional OPTIONS to pass to ADB
  -o, --nunit-options=OPTIONS
                             Additional OPTIONS to pass to NUnit console runner
```

So yes, they are additive and, no, they have no effect when running a group.

>  -i, --include-categories=CATEGORIES
>                             Run tests in the specified CATEGORIES. Takes a
>                               comma-separated list of test categories
> So tests have categories? (This is presumably the NUnit/xUnit concept of categories?) How does this relate to groups & suites? Does it? Should it?

Some do. Some don't. It shouldn't relate to groups, see the previous explanation of groups.

OK, this is getting way to long to follow so let me reply shortly to the rest of your review

  * Regarding test suite names: they are generated from the paths and project names to minimize the maintenance effort. Introducing "display aliases", "tags" or whatever other criteria can be done.
  * I'm not sure why "Host unit" isn't clear? We're dealing with tests and we all know what host and unit tests are?
  * Output is verbose because this is Work In Progress and I **need** it to work more comfortably - this, and the help screen wording etc, are the last things I hoped to focus on after all tests run properly and groups (or whatever name we end up using) are decided upon. I can't make the decision myself because the topic is so vast - hence the request to ALL of XA team to review and suggest the groupings.

Jon, thanks for the review but I **still** don't know how to move on with the code. You ask a lot of questions, but there are few answers I need to move on. I'll add a separate comment regarding the input I need right now to keep working on this.

**Comment by grendello** [comment]
@jonpryor maybe this graphics (I'm not an artist :P) will help in visualizing what I had in mind for groups/suites etc etc

groups

**Comment by grendello** [comment]
This isn't likely to be merged, since we'll probably migrate to `dotnet test`, closing.

**Pull request closed without merging**